File: primes/primes.py
import os
import platform
import socket
import sys

UPPER_BOUND = 5000000
PREFIX = 32338


class Node:
    def __init__(self):
        self.children = {}
        self.terminal = False


class Sieve:
    def __init__(self, limit):
        self.limit = limit
        self.prime = [False] * (limit + 1)

    def to_list(self):
        result = [2, 3]
        for p in range(5, self.limit + 1):
            if self.prime[p]:
                result.append(p)
        return result

    def omit_squares(self):
        r = 5
        while r * r < self.limit:
            if self.prime[r]:
                i = r * r
                while i < self.limit:
                    self.prime[i] = False
                    i = i + r * r
            r = r + 1
        return self

    def step1(self, x, y):
        n = (4 * x * x) + (y * y)
        if n <= self.limit and (n % 12 == 1 or n % 12 == 5):
            self.prime[n] = not self.prime[n]

    def step2(self, x, y):
        n = (3 * x * x) + (y * y)
        if n <= self.limit and n % 12 == 7:
            self.prime[n] = not self.prime[n]

    def step3(self, x, y):
        n = (3 * x * x) - (y * y)
        if x > y and n <= self.limit and n % 12 == 11:
            self.prime[n] = not self.prime[n]

    def loop_y(self, x):
        y = 1
        while y * y < self.limit:
            self.step1(x, y)
            self.step2(x, y)
            self.step3(x, y)
            y = y + 1

    def loop_x(self):
        x = 1
        while x * x < self.limit:
            self.loop_y(x)
            x = x + 1

    def calc(self):
        self.loop_x()
        return self.omit_squares()


def generate_trie(l):
    root = Node()
    for el in l:
        head = root
        for ch in str(el):
            if ch not in head.children:
                head.children[ch] = Node()
            head = head.children[ch]
        head.terminal = True
    return root


def find(upper_bound, prefix):
    primes = Sieve(upper_bound).calc()
    str_prefix = str(prefix)
    head = generate_trie(primes.to_list())
    for ch in str_prefix:
        head = head.children.get(ch)
        if head is None:
            return None

    queue, result = [(head, str_prefix)], []
    while queue:
        top, prefix = queue.pop()
        if top.terminal:
            result.append(int(prefix))
        for ch, v in top.children.items():
            queue.insert(0, (v, prefix + ch))
    return result


def notify(msg):
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
        if not s.connect_ex(("localhost", 9001)):
            s.sendall(bytes(msg, "utf8"))


def verify():
    left = [2, 23, 29]
    right = find(100, 2)
    if left != right:
        print("%s != %s" % (left, right), file=sys.stderr)
        quit(1)


if __name__ == "__main__":
    verify()

    notify("%s\t%d" % (platform.python_implementation(), os.getpid()))
    results = find(UPPER_BOUND, PREFIX)
    notify("stop")

    print(results)

File: primes/primes.jl
using Sockets


const UPPER_BOUND = 5_000_000
const PREFIX = 32_338


mutable struct Node
    children::Dict{Char,Node}
    terminal::Bool
end
Node(children::Dict) = Node(children, false)
Node() = Node(Dict(), false)


mutable struct Sieve
    limit::Int64
    prime::BitVector

    function Sieve(limit::Int64)
        return new(limit, falses(limit + 1))
    end
end


function to_list(s::Sieve)::Vector{Int64}
    result = [2, 3]
    @inbounds for p = 1:s.limit+1
        if s.prime[p] == true
            push!(result, p)
        end
    end
    return result
end


function omit_squares!(s::Sieve)
    r = 5
    while r^2 < s.limit
        @inbounds if s.prime[r]
            i = r^2
            while i < s.limit
                s.prime[i] = false
                i += r^2
            end
        end
        r += 1
    end
end


function step1!(s::Sieve, x::Int64, y::Int64)
    n = 4x^2 + y^2
    if n <= s.limit && (mod(n, 12) in (1, 5))
        @inbounds s.prime[n] = !s.prime[n]
    end
end


function step2!(s::Sieve, x::Int64, y::Int64)
    n = 3x^2 + y^2
    if n <= s.limit && mod(n, 12) == 7
        @inbounds s.prime[n] = !s.prime[n]
    end
end


function step3!(s::Sieve, x::Int64, y::Int64)
    n = 3x^2 - y^2
    if x > y && n <= s.limit && mod(n, 12) == 11
        @inbounds s.prime[n] = !s.prime[n]
    end
end


function loop_y!(s::Sieve, x::Int64)
    y = 1
    while y^2 < s.limit
        step1!(s, x, y)
        step2!(s, x, y)
        step3!(s, x, y)
        y += 1
    end
end


function loop_x!(s::Sieve)
    x = 1
    while x^2 < s.limit
        loop_y!(s, x)
        x += 1
    end
end


function calc(s::Sieve)
    loop_x!(s)
    omit_squares!(s)
    return s
end


function generate_trie(prime_list::Vector{Int64})::Node
    root = Node()
    for p in prime_list
        head = root
        for ch in string(p)
            if !(ch in keys(head.children))
                head.children[ch] = Node()
            end
            head = head.children[ch]
        end
        head.terminal = true
    end
    return root
end


function find(upper_bound::Int64, search_prefix::Int64)::Vector
    primes = calc(Sieve(upper_bound))
    str_prefix = string(search_prefix)

    head = generate_trie(to_list(primes))

    for char in str_prefix
        head = get(head.children, char, nothing)
        head == nothing && return nothing
    end

    queue = [(head, str_prefix)]
    result = Int64[]

    while !isempty(queue)
        top, prefix = pop!(queue)

        if top.terminal == true
            push!(result, parse(Int64, prefix))
        end

        for (char, v) in top.children
            pushfirst!(queue, (v, prefix * char))
        end
    end

    sort!(result)
    return result
end


function notify(msg)
    try
        socket = connect("localhost", 9001)
        write(socket, msg)
        close(socket)
    catch
        # standalone usage
    end
end


function verify()
    left = [2, 23, 29]
    right = find(100, 2)

    if left != right
        println("$left != $right")
        exit()
    end
end


if abspath(PROGRAM_FILE) == @__FILE__
    verify()

    notify("Julia\t$(getpid())")
    results = find(UPPER_BOUND, PREFIX)
    notify("stop")

    println(results)
end
